### PR TITLE
[BUG]: Shift box_aspect according to vertical_axis

### LINKED ISSUE #24328: [Bug]: class Axes3D.set_box_aspect() sets wrong aspect ratios when Axes3D.view_init( vertical_axis='y') is enabled.
### Bug summary

Whenever creating a 3D plot with `vertical_axis = 'y'` option the `set_box_aspect()` method scales the axes incorrectly. The 3-tuple is expected to behave as `x:y:z`, instead it behaves like `z:x:y`

### Code for reproduction

```python
"""
Created on Mon Oct 31 18:30:31 2022

@author: Michel Gordillo
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm

fig = plt.figure(num=4, clear=True)
ax = fig.add_subplot(1, 1, 1, projection='3d')
ax.view_init(vertical_axis='y')

(theta, phi) = np.meshgrid(np.linspace(0, 2 * np.pi, 41),
                           np.linspace(0, 2 * np.pi, 41))

x = (3 + np.cos(phi)) * np.cos(theta)
z = (3 + np.cos(phi)) * np.sin(theta)
y = np.sin(phi)

k = 4

xlim = [-k, k]  #2k
ylim = [-k/2, k/2] #k
zlim = [-k, k] #2k

box_aspect = (4,2,4) #(x,y,z)

def fun(t, f): return (np.cos(f + 2 * t) + 1) / 2


dplot = ax.plot_surface(x, y, z, facecolors=cm.jet(fun(theta, phi)))
ax.set(xlabel='x', 
       ylabel='y', 
       zlabel='z',
       xlim = xlim,
       ylim = ylim,
       zlim = zlim,
       # xticks = [-4, -2, 2, 4],
       # yticks = [-4, -2, 2, 4],
       # zticks = [-1, 0, 1],
       title='Donut!')

ax.set_box_aspect(aspect = box_aspect)

fig.tight_layout()
```

### Actual outcome

Matplotlib_bug_axes_vertical_y

The box_aspect tuple:
`box_aspect = (4,2,4)` #(x,y,z)

Actually the axis scaled as
`box_aspect = (4,2,4)`  -> #(z,x,y) 
Like if shifting y to the right.

### Expected outcome

Matplotlib_bug_fix_axes_vertical_y

To produce this result i just applied the same logic in reverse:
`box_aspect = (4,4,2)` #(z,x,y) <- Expecting the outcome to be applied like this

And it actually does. 

### Additional information

This only happens as explained in the title when both `vertical = y` option is enabled and we try to manually scale_box_aspect(). The problem is probably from how was vertical axis selection actually implemented.

From the documentation: 

> set_box_aspect(aspect, *, zoom=1) 
[...]
aspect3-tuple of floats or None
Changes the physical dimensions of the Axes3D, such that the ratio of the axis lengths in display units is x:y:z.

This behaviour is not respected when `vertical_axis = 'y'` in enabled. The behaviour appears to be: `z:x:y` instead.

To actually fix this I just created a tuple with the new definition in mind, this could just be patched by shifting the tuple to make the vertical axis be the last element of the tuple

 (x,y,z) -> input tuple

vertical axis == 'z' -> (x,y,z) no problem
vertical axis == 'y' -> (z,x,y) rotate right the tuple 
vertical axis == 'x' -> (y,z,x) rotate right twice the tuple 

### Operating system

Windows

### Matplotlib Version

'3.6.1'

### Matplotlib Backend

Qt5Agg

### Python version

Python 3.8.5

### Jupyter version

Spyder 4.1.5

### Installation

conda

### LINKED ISSUE #28052: [Bug]: set_aspect incompatible with view_init
### Bug summary

ax.set_aspect('equal') doesn't work properly on axes where the view has been manipulated, i.e.  ax.view_init(vertical_axis='y').
It sets the aspect ratio as if the axes were still oriented in the standard way. 

### Code for reproduction

```Python
import numpy as np
import matplotlib.pyplot as plt
import mpl_toolkits.mplot3d as a3

plt.close('all')
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.view_init(vertical_axis='y')
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_xlabel('Z')
ax.axes.set_xlim3d(left=-1, right=3.5) 
ax.axes.set_ylim3d(bottom=0, top=10) 
ax.axes.set_zlim3d(bottom=-0.5, top=3.5) 
ax.set_aspect('equal')
    
def a(elemList):
    return np.array(elemList)

def drawTriangle(centre, size=1, color='y'):
    # A neuron is a triangle
    left = centre + a([- size / 2, - size / 2, 0])
    right = centre + a([size / 2, - size / 2, 0])
    top = centre + a([0, size / 2, 0])
    coords = a([left, top, right])
    tri = a3.art3d.Poly3DCollection([coords])
    tri.set_color(color)
    tri.set_edgecolor('k')
    ax.add_collection3d(tri)

for col in range(3):
    drawTriangle(a([col, 10, 0]))
```

### Actual outcome

image

### Expected outcome

You would want the vertical 'y' axis to be stretched to 10 units, instead the horizontal 'z' axis has been stretched.

### Additional information

_No response_

### Operating system

_No response_

### Matplotlib Version

3.7.2

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Installation

None

### PR DESCRIPTION
**Pull request by Illviljan**
## PR summary
Make sure box_aspect stays the same for each axis when using `set_box_aspect` and vertical_axis is used.

```python
import matplotlib.pyplot as plt
import numpy as np


x = np.array([0, 1, 2, 4])
y = np.array([5, 10])
z = np.array([100, 150, 200])
X, Y, Z = np.meshgrid(*[x, y, z], indexing="ij")
c = np.arange(0, X.size)

plt.figure()
for j, (a, e) in enumerate([(0, 0), (-60, 30)]):
    for i, vert_a in enumerate(["z", "y", "x"]):
        ax = plt.subplot(2, 3, j * 3 + 1 + i, projection="3d")
        pc = ax.scatter(X, Y, Z, c=c)
        ax.view_init(azim=a, elev=e, vertical_axis=vert_a)
        ax.set_title(f"azim={a}, elev={e}, vertical_axis='{vert_a}'")
        ax.set_xlabel("x")
        ax.set_ylabel("y")
        ax.set_zlabel("z")
        ax.set_box_aspect((1, 2, 3))
```

Before PR:
image

With PR:
image

## PR checklist

- [x] Closes #24328
- [x] Closes #28052
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)

### CODE REVIEW

**Comment by Illviljan** [comment]
Fixes set_aspect as well, since it uses set_box_aspect.

<details>

```
import numpy as np
import matplotlib.pyplot as plt
import mpl_toolkits.mplot3d as a3

plt.close('all')
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.view_init(vertical_axis='y')
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_xlabel('Z')
ax.axes.set_xlim3d(left=-1, right=3.5) 
ax.axes.set_ylim3d(bottom=0, top=10) 
ax.axes.set_zlim3d(bottom=-0.5, top=3.5) 
ax.set_aspect('equal')
    
def a(elemList):
    return np.array(elemList)

def drawTriangle(centre, size=1, color='y'):
    # A neuron is a triangle
    left = centre + a([- size / 2, - size / 2, 0])
    right = centre + a([size / 2, - size / 2, 0])
    top = centre + a([0, size / 2, 0])
    coords = a([left, top, right])
    tri = a3.art3d.Poly3DCollection([coords])
    tri.set_color(color)
    tri.set_edgecolor('k')
    ax.add_collection3d(tri)

for col in range(3):
    drawTriangle(a([col, 10, 0]))
```
</details>

With PR:
image

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -383,7 +383,8 @@ def set_box_aspect(self, aspect, *, zoom=1):
         # of the axes in mpl3.8.
         aspect *= 1.8294640721620434 * 25/24 * zoom / np.linalg.norm(aspect)
 
-        self._box_aspect = aspect
+        self._box_aspect = self._roll_to_vertical(aspect, sign=-1)
+        # self._box_aspect = aspect
```
Comment (scottshambaugh) [inline lib/mpl_toolkits/mplot3d/axes3d.py:387]:
```suggestion
```

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 2]
```diff
@@ -1190,15 +1191,16 @@ def set_proj_type(self, proj_type, focal_length=None):
...
         """Roll arrays to match the different vertical axis."""
-        return np.roll(arr, self._vertical_axis - 2)
+        return np.roll(arr, sign * (self._vertical_axis - 2))
 
     def get_proj(self):
         """Create the projection matrix from the current viewing position."""
 
         # Transform to uniform world coordinates 0-1, 0-1, 0-1
         box_aspect = self._roll_to_vertical(self._box_aspect)
+        # self._box_aspect = box_aspect
```
Comment (scottshambaugh) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1203]: What's the reason for removing this line? I'm not totally clear where it comes into play (when rotating the view?), but note that the 3D test coverage isn't amazing so we should understand where behavior is changing.

**Review by scottshambaugh — COMMENTED** [review]
Thanks for fixing this as well! Two small notes, then I can approve the MR.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 2 above)
Comment (Illviljan) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1203]: It was never there in main. I was just testing different ideas here.

Conclusion here is that box_aspect has to be local, otherwise there will be a new vertical axis each click.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 3]
```diff
@@ -1190,15 +1190,16 @@ def set_proj_type(self, proj_type, focal_length=None):
...
```
Comment (Illviljan) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1202]:
```suggestion
```

**Review by scottshambaugh — APPROVED** [review]
CI failure looks unrelated to me, good to merge!

**Comment by scottshambaugh** [comment]
It looks like this needs a rebase to deconflict with your other MR that was just merged in.

File: lib/mpl_toolkits/mplot3d/tests/test_axes3d.py [hunk 4]
```diff
@@ -2276,6 +2276,23 @@ def test_on_move_vertical_axis(vertical_axis: str) -> None:
...
+
+    aspect_old = tuple(ax._box_aspect)
+    aspect_expected = np.roll(
+        aspect_old, -1 * (ax._axis_names.index(vertical_axis) - 2)
+    )
+
+    ax.set_box_aspect(None)
+    aspect_new = tuple(ax._box_aspect)
+
+    np.testing.assert_allclose(aspect_expected, aspect_new)
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:2293]: Does this test have semantic meaning? I can't see by itself that `aspect_expected` is expected to be as defined by the roll operation. To me it appears that this formular was just copied from `set_box_aspect`, which would be tautological ("set_box_aspect behaves according to the formula given in set_box_aspect") and doesn't actually prove anything. Also note that the first two entries of the default box aspect (None -> (4, 4, 3)) are identical. This reduces test power because we cannot distinguish whether they are mixed up or not.

I think there are two ways to make this better: Either

- Use explicit expectations. Something like (not checked the numbers and the exact logic):
  ```
  ax.view_init(elev=0, azim=0, roll=0, vertical_axis="x")
  ax.set_box_aspect((1, 2, 3))  # (x, y, z)
  np.testing.assert_allclose)(ax._box_aspect, (2, 3, 1))  # (width, depth, height)
  ```
  or
- use an image comparison test and compare rotated views with permuted box_aspects.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 5]
```diff
@@ -1191,9 +1192,21 @@ def set_proj_type(self, proj_type, focal_length=None):
...
+        """
+        Roll arrays to match the different vertical axis.
+
+        Parameters
+        ----------
+        arr : ArrayLike
+            Array to roll.
+        sign : Literal[1, -1], default: 1
+            Roll the array elements in a positive or negative
+            direction. Defaults to the positive direction.
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1207]: My point was that the naming `sign` itself is not good because it's on a technical level not reflecting the underlying semantics. Expanding the documentation of the technical action only marginally improves understandability. To be explicit: Please change the name.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 6]
```diff
@@ -1191,9 +1192,21 @@ def set_proj_type(self, proj_type, focal_length=None):
                                  f"None for proj_type = {proj_type}")
             self._focal_length = np.inf
 
-    def _roll_to_vertical(self, arr):
-        """Roll arrays to match the different vertical axis."""
-        return np.roll(arr, self._vertical_axis - 2)
+    def _roll_to_vertical(
+        self, arr: "np.typing.ArrayLike", sign: Literal[1, -1] = 1
+    ) -> np.ndarray:
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1197]: @QuLogic it seems `mplot3d` is not yet typed at all. Do you plan to go with type stubs or inline type hints here?

Comment (ksunden) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1197]: Correct that all of mpltoolkits (including mplot3d) is not yet type hinted at all.

I have no current plans regarding extending type hints here at the moment (not opposed, just have no active plans to undertake it myself)

I think whether to do stubs or inline is up to whoever wishes to implement it. Stubs would be my default option, just because that is what we have mostly, but it is not a strong opinion.

I _do_ prefer inline in general, but opted for stubs because it limited how much runtime code needed to be touched in an already pretty large task.

That said, we have precedence for some new typing work being done inline (`_mathtext.py`).

Further, I have in general erred towards not type hinting private functionality (as this is), at least in stubs, as they are mostly useful to external users anyway. 


TL;DR Could go either way, not really harming anything to have them here.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 5 above)
Comment (Illviljan) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: sign was inspired by [np.sign](https://numpy.org/doc/stable/reference/generated/numpy.sign.html) for reference. 
The arguments I see against `reverse: bool` is that it leads to more code complexity (an additional if-check). This function is triggered every time on move so some performance consideration has to be made in my opinion. Probably not the worst bottleneck though.

**Comment by Illviljan** [comment]
Explicit test values:
<details>

```python
vertical_axis = 'x', aspect_expected = [1.14285715, 0.85714286, 1.14285715]

    @pytest.mark.parametrize(
        "vertical_axis, aspect_expected",
        [
            ("x", [1.14285715, 0.85714286, 1.14285715]),
            ("y", [0.85714286, 1.14285715, 1.14285715]),
            ("z", [1.14285715, 1.14285715, 0.85714286]),
        ],
    )
    def test_set_box_aspect_vertical_axis(vertical_axis, aspect_expected):
        ax = plt.subplot(1, 1, 1, projection="3d")
        ax.view_init(elev=0, azim=0, roll=0, vertical_axis=vertical_axis)
        ax.figure.canvas.draw()
    
        ax.set_box_aspect(None)
    
>       np.testing.assert_allclose(aspect_expected, ax._box_aspect)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 3 / 3 (100%)
E       Max absolute difference: 0.04761904
E       Max relative difference: 0.04
E        x: array([1.142857, 0.857143, 1.142857])
E        y: array([1.190476, 0.892857, 1.190476])


vertical_axis = 'y', aspect_expected = [0.85714286, 1.14285715, 1.14285715]

    @pytest.mark.parametrize(
        "vertical_axis, aspect_expected",
        [
            ("x", [1.14285715, 0.85714286, 1.14285715]),
            ("y", [0.85714286, 1.14285715, 1.14285715]),
            ("z", [1.14285715, 1.14285715, 0.85714286]),
        ],
    )
    def test_set_box_aspect_vertical_axis(vertical_axis, aspect_expected):
        ax = plt.subplot(1, 1, 1, projection="3d")
        ax.view_init(elev=0, azim=0, roll=0, vertical_axis=vertical_axis)
        ax.figure.canvas.draw()
    
        ax.set_box_aspect(None)
    
>       np.testing.assert_allclose(aspect_expected, ax._box_aspect)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 3 / 3 (100%)
E       Max absolute difference: 0.04761904
E       Max relative difference: 0.04
E        x: array([0.857143, 1.142857, 1.142857])
E        y: array([0.892857, 1.190476, 1.190476])


 vertical_axis = 'z', aspect_expected = [1.14285715, 1.14285715, 0.85714286]

    @pytest.mark.parametrize(
        "vertical_axis, aspect_expected",
        [
            ("x", [1.14285715, 0.85714286, 1.14285715]),
            ("y", [0.85714286, 1.14285715, 1.14285715]),
            ("z", [1.14285715, 1.14285715, 0.85714286]),
        ],
    )
    def test_set_box_aspect_vertical_axis(vertical_axis, aspect_expected):
        ax = plt.subplot(1, 1, 1, projection="3d")
        ax.view_init(elev=0, azim=0, roll=0, vertical_axis=vertical_axis)
        ax.figure.canvas.draw()
    
        ax.set_box_aspect(None)
    
>       np.testing.assert_allclose(aspect_expected, ax._box_aspect)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 3 / 3 (100%)
E       Max absolute difference: 0.04761904
E       Max relative difference: 0.04
E        x: array([1.142857, 1.142857, 0.857143])
E        y: array([1.190476, 1.190476, 0.892857])
```

</details>

**Review by timhoffm — APPROVED** [review]

File: lib/mpl_toolkits/mplot3d/tests/test_axes3d.py [hunk 7]
```diff
@@ -2276,6 +2276,24 @@ def test_on_move_vertical_axis(vertical_axis: str) -> None:
...
+    ],
+)
+def test_set_box_aspect_vertical_axis(vertical_axis, aspect_expected):
+    ax = plt.subplot(1, 1, 1, projection="3d")
+    ax.view_init(elev=0, azim=0, roll=0, vertical_axis=vertical_axis)
+    ax.figure.canvas.draw()
+
+    ax.set_box_aspect(None)
+
+    np.testing.assert_allclose(aspect_expected, ax._box_aspect)
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:2294]:
```suggestion
    np.testing.assert_allclose(aspect_expected, ax._box_aspect, rtol=1e-6)
```
Should fix the issue.
